src/vulkan-renderer/world/collision.cpp

```diff
@@ -0,0 +1,184 @@
+    m_nearest_edge = std::move(other.m_nearest_edge);
+}
+
+template <typename T>
+RayCubeCollision<T>::RayCubeCollision(const T &cube, const glm::vec3 ray_pos, const glm::vec3 ray_dir) : m_cube(cube) {
+
+    // This lambda adjusts the center points on a cube's face to the size of the octree,
+    // so collision works with cubes of any size. This does not yet account for rotations!
+    const auto adjust_coordinates = [=](const glm::vec3 pos) {
+        // TODO: Take rotation of the cube into account.
+        return m_cube.center() + pos * (m_cube.size() / 2);
+    };
+
+    const std::array cube_directions{
+        glm::vec3(-1.0f, 0.0f, 0.0f), // left
+        glm::vec3(1.0f, 0.0f, 0.0f),  // right
+        glm::vec3(0.0f, -1.0f, 0.0f), // front
+        glm::vec3(0.0f, 1.0f, 0.0f),  // back
+        glm::vec3(0.0f, 0.0f, 1.0f),  // top
+        glm::vec3(0.0f, 0.0f, -1.0f)  // bottom
+    };
+
+    // The coordinates of the center of every face of the cube.
+    const std::array cube_face_centers{cube_face{"left", adjust_coordinates(cube_directions[0])},
+                                       cube_face{"right", adjust_coordinates(cube_directions[1])},
+                                       cube_face{"front", adjust_coordinates(cube_directions[2])},
+                                       cube_face{"back", adjust_coordinates(cube_directions[3])},
+                                       cube_face{"top", adjust_coordinates(cube_directions[4])},
+                                       cube_face{"bottom", adjust_coordinates(cube_directions[5])}};
+
+    // The coordinates of every corner of the cube.
+    std::array cube_corners = {cube_corner{"left back top", adjust_coordinates({-1.0f, 1.0f, 1.0f})},
+                               cube_corner{"right back top", adjust_coordinates({1.0f, 1.0f, 1.0f})},
+                               cube_corner{"left back botom", adjust_coordinates({-1.0f, 1.0f, -1.0f})},
+                               cube_corner{"right back bottom", adjust_coordinates({1.0f, 1.0f, -1.0f})},
+                               cube_corner{"left front top", adjust_coordinates({-1.0f, -1.0f, 1.0f})},
+                               cube_corner{"right front top", adjust_coordinates({1.0f, -1.0f, 1.0f})},
+                               cube_corner{"left front bottom", adjust_coordinates({-1.0f, -1.0f, -1.0f})},
+                               cube_corner{"right front bottom", adjust_coordinates({1.0f, -1.0f, -1.0f})}};
+
+    using corner_on_face_index = std::array<std::size_t, 4>;
+
+    // These indices specify which 4 points describe the corners on the given face.
+    static constexpr std::array CORNERS_ON_FACE_INDICES{
+        corner_on_face_index{0, 2, 4, 6}, // left
+        corner_on_face_index{1, 3, 5, 7}, // right
+        corner_on_face_index{4, 5, 6, 6}, // front
+        corner_on_face_index{0, 1, 2, 3}, // back
+        corner_on_face_index{0, 1, 4, 5}, // top
+        corner_on_face_index{2, 3, 6, 7}  // bottom
```

Comment:
Does this match with our current corner ids?

Reply:
No, it must be:

```cpp
left: 0,1,2,3
right: 4,5,6,7
front: 0,1,4,5
back: 2,3,6,7
top: 1,3,5,7
bottom: 0,2,4,6
```

Reply:
Done.